rig-lancedb/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+}
+
+/// See [LanceDB vector search](https://lancedb.github.io/lancedb/search/) for more information.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub enum SearchType {
+    // Flat search, also called ENN or kNN.
+    Flat,
+    /// Approximal Nearest Neighbor search, also called ANN.
+    Approximate,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+pub struct SearchParams {
+    /// Always set the distance_type to match the value used to train the index
+    /// By default, set to L2
+    distance_type: Option<DistanceType>,
+    /// By default, ANN will be used if there is an index on the table.
+    /// By default, kNN will be used if there is NO index on the table.
+    /// To use defaults, set to None.
+    search_type: Option<SearchType>,
+    /// Set this value only when search type is ANN.
+    /// See [LanceDb ANN Search](https://lancedb.github.io/lancedb/ann_indexes/#querying-an-ann-index) for more information
+    nprobes: Option<usize>,
+    /// Set this value only when search type is ANN.
+    /// See [LanceDb ANN Search](https://lancedb.github.io/lancedb/ann_indexes/#querying-an-ann-index) for more information
+    refine_factor: Option<u32>,
+    /// If set to true, filtering will happen after the vector search instead of before
+    /// See [LanceDb pre/post filtering](https://lancedb.github.io/lancedb/sql/#pre-and-post-filtering) for more information
+    post_filter: Option<bool>,
+}
+
+impl SearchParams {
+    pub fn distance_type(mut self, distance_type: DistanceType) -> Self {
+        self.distance_type = Some(distance_type);
+        self
+    }
+
+    pub fn search_type(mut self, search_type: SearchType) -> Self {
+        self.search_type = Some(search_type);
+        self
+    }
+
+    pub fn nprobes(mut self, nprobes: usize) -> Self {
+        self.nprobes = Some(nprobes);
+        self
+    }
+
+    pub fn refine_factor(mut self, refine_factor: u32) -> Self {
+        self.refine_factor = Some(refine_factor);
+        self
+    }
+
+    pub fn post_filter(mut self, post_filter: bool) -> Self {
+        self.post_filter = Some(post_filter);
+        self
+    }
+}
+
+impl<M: EmbeddingModel> LanceDbVectorStore<M> {
+    /// Note: Tables are created inside the new function rather than created outside and passed as reference to new function.
+    /// This is because a specific schema needs to be enforced on the tables and this is done at creation time.
+    pub async fn new(
+        db: &lancedb::Connection,
+        model: &M,
+        search_params: &SearchParams,
+    ) -> Result<Self, lancedb::Error> {
+        let document_table = db
+            .create_empty_table("documents", Arc::new(Self::document_schema()))
+            .execute()
+            .await?;
+
+        let embedding_table = db
+            .create_empty_table(
+                "embeddings",
+                Arc::new(Self::embedding_schema(model.ndims() as i32)),
+            )
+            .execute()
+            .await?;
```

Comment:
Either make the tables arguments of the constructor, or add a check to see if they exist. Rn, this will fail if the tables already exist (e.g.: when running the same example twice in a row)